File: backend/services/file-storage-service/src/main/java/com/lmspilot/filestorage/api/FileStorageApi.java
package com.lmspilot.filestorage.api;

import com.lmspilot.filestorage.domain.*;

import com.lmspilot.support.api.ApiException;

import com.lmspilot.support.security.CurrentUser;

import jakarta.annotation.PostConstruct;

import java.io.*;

import java.nio.charset.StandardCharsets;

import java.nio.file.*;

import java.security.*;

import java.time.*;

import java.util.*;

import java.util.zip.*;

import javax.xml.parsers.*;

import org.springframework.beans.factory.annotation.Value;

import org.springframework.core.io.*;

import org.springframework.http.*;

import org.springframework.stereotype.Service;

import org.springframework.transaction.annotation.Transactional;

import org.springframework.web.bind.annotation.*;

import org.springframework.web.multipart.MultipartFile;

import org.w3c.dom.*;
record StoredFileResponse(UUID id,String originalName,String contentType,long sizeBytes,String sha256,String purpose,StoredFileStatus status,Instant createdAt){
}
record DocumentPreviewResponse(UUID fileId,String originalName,List<String> paragraphs){
}
record InternalStoredFileResponse(UUID id,UUID ownerId,String originalName,String contentType,long sizeBytes,String sha256,String purpose,StoredFileStatus status,Instant createdAt){
}
record GrantFileAccessRequest(UUID fileId,UUID userId,String source,Instant expiresAt){
}
@Service
@Transactional
class FileStorageService {
    private final StoredFileRepository repo;
    private final FileAccessGrantRepository grants;
    private final FileVersionRepository versions;
    private final Path root;
    private final long max;
    private final Set<String> allowed;
    private static final Set<String> BLOCKED=Set.of("exe","dll","bat","cmd","ps1","sh","jar","msi","com","scr","js","html","htm","svg");
    public FileStorageService(StoredFileRepository r,FileAccessGrantRepository g,FileVersionRepository v,@Value("${storage.root:./data/files}")String root,@Value("${storage.max-size-bytes:209715200}")long max,@Value("${storage.allowed-content-types:application/pdf,application/vnd.openxmlformats-officedocument.wordprocessingml.document,video/mp4,audio/mpeg,image/png,image/jpeg,image/webp,text/plain,text/csv}")String types){
        repo=r;
        grants=g;
        versions=v;
        this.root=Paths.get(root).toAbsolutePath().normalize();
        this.max=max;
        allowed=new HashSet<>(Arrays.asList(types.toLowerCase(Locale.ROOT).split(",")));
    }
    @PostConstruct
    public void init()throws IOException{
        Files.createDirectories(root);
        if(max<=0)throw new IllegalStateException("storage.max-size-bytes must be positive");
    }
    StoredFileResponse store(MultipartFile file,String purpose){
        if(file==null||file.isEmpty())bad("FILE_EMPTY","Tệp tải lên đang trống");
        if(file.getSize()>max)throw new ApiException(HttpStatus.PAYLOAD_TOO_LARGE,"FILE_TOO_LARGE","Tệp vượt quá dung lượng cho phép");
        String name=safe(file.getOriginalFilename());
        String ext=name.contains(".")?name.substring(name.lastIndexOf('.')+1).toLowerCase(Locale.ROOT):"";
        if(ext.isBlank()||BLOCKED.contains(ext))throw new ApiException(HttpStatus.UNSUPPORTED_MEDIA_TYPE,"FILE_TYPE_BLOCKED","Loại tệp bị chặn");
        String content=file.getContentType()==null?"application/octet-stream":file.getContentType().toLowerCase(Locale.ROOT);
        if(!allowed.contains(content))throw new ApiException(HttpStatus.UNSUPPORTED_MEDIA_TYPE,"FILE_TYPE_NOT_ALLOWED","Loại nội dung không được cho phép");
        String pur=(purpose==null?"GENERAL":purpose.trim().toUpperCase(Locale.ROOT));
        if(pur.length()>50)bad("PURPOSE_INVALID","Mục đích tệp quá dài");
        UUID id=UUID.randomUUID();
        String key=id.toString().substring(0,2)+"/"+id;
        Path target=path(key);
        try{
            Files.createDirectories(target.getParent());
            byte[] bytes=file.getBytes();
            Files.write(target,bytes,StandardOpenOption.CREATE_NEW);
            StoredFileEntity e=new StoredFileEntity();
            e.id=id;
            e.ownerId=user();
            e.originalName=name;
            e.storageKey=key;
            e.contentType=content;
            e.sizeBytes=bytes.length;
            e.sha256=sha(bytes);
            e.purpose=pur;
            repo.save(e);
            FileVersionEntity ver=new FileVersionEntity();
            ver.fileId=e.id;
            ver.storageKey=key;
            ver.mediaType=content;
            ver.sizeBytes=e.sizeBytes;
            ver.sha256=e.sha256;
            ver.createdBy=e.ownerId;
            versions.save(ver);
            return view(e);
        }
        catch(IOException x){
            throw new ApiException(HttpStatus.INTERNAL_SERVER_ERROR,"FILE_STORE_FAILED","Không thể lưu tệp");
        }

    }
    List<StoredFileResponse> list(){
        return repo.findAllByOwnerIdOrderByCreatedAtDesc(user()).stream().filter(e->e.status!=StoredFileStatus.DELETED).map(this::view).toList();
    }
    StoredFileResponse metadata(UUID id){
        return view(readable(id));
    }
    InternalStoredFileResponse internalMetadata(UUID id){
        StoredFileEntity e=require(id);
        return new InternalStoredFileResponse(e.id,e.ownerId,e.originalName,e.contentType,e.sizeBytes,e.sha256,e.purpose,e.status,e.createdAt);
    }
    ResponseEntity<Resource> content(UUID id,String range){
        StoredFileEntity e=readable(id);
        return resource(e);
    }
    ResponseEntity<Resource> internalContent(UUID id){
        return resource(require(id));
    }
    DocumentPreviewResponse preview(UUID id){
        StoredFileEntity e=readable(id);
        if(!e.contentType.contains("wordprocessingml"))throw new ApiException(HttpStatus.UNSUPPORTED_MEDIA_TYPE,"DOCX_REQUIRED","Tệp không phải DOCX");
        List<String> p=new ArrayList<>();
        try(ZipFile z=new ZipFile(path(e.storageKey).toFile())){
            ZipEntry entry=z.getEntry("word/document.xml");
            if(entry==null)return new DocumentPreviewResponse(id,e.originalName,List.of());
            DocumentBuilderFactory f=DocumentBuilderFactory.newInstance();
            f.setFeature("http://apache.org/xml/features/disallow-doctype-decl",true);
            f.setFeature("http://xml.org/sax/features/external-general-entities",false);
            Document d=f.newDocumentBuilder().parse(z.getInputStream(entry));
            NodeList nodes=d.getElementsByTagNameNS("*","p");
            for(int i=0;
            i<nodes.getLength();
            i++){
                String s=nodes.item(i).getTextContent().trim();
                if(!s.isBlank())p.add(s);
            }

        }
        catch(Exception x){
            throw new ApiException(HttpStatus.UNPROCESSABLE_ENTITY,"DOCX_PREVIEW_FAILED","Không thể đọc nội dung DOCX");
        }
        return new DocumentPreviewResponse(id,e.originalName,p);
    }
    void delete(UUID id){
        StoredFileEntity e=readable(id);
        e.status=StoredFileStatus.DELETED;
        e.deletedAt=Instant.now();
        repo.save(e);
    }
    void grant(GrantFileAccessRequest i){
        FileAccessGrantEntity g=grants.findByFileIdAndUserId(i.fileId(),i.userId()).orElseGet(FileAccessGrantEntity::new);
        g.fileId=i.fileId();
        g.userId=i.userId();
        g.source=i.source()==null?"INTERNAL":i.source();
        g.expiresAt=i.expiresAt()==null?Instant.now().plus(Duration.ofDays(30)):i.expiresAt();
        g.updatedAt=Instant.now();
        grants.save(g);
    }
    StoredFileEntity require(UUID id){
        return repo.findById(id).filter(e->e.status==StoredFileStatus.AVAILABLE).orElseThrow(()->new ApiException(HttpStatus.NOT_FOUND,"FILE_NOT_FOUND","Không tìm thấy tệp"));
    }
    Path filePath(StoredFileEntity e){
        return path(e.storageKey);
    }
    private StoredFileEntity readable(UUID id){
        StoredFileEntity e=require(id);
        UUID u=user();
        if(!e.ownerId.equals(u)){
            boolean ok=grants.findByFileIdAndUserId(id,u).filter(g->g.expiresAt.isAfter(Instant.now())).isPresent();
            if(!ok)throw new ApiException(HttpStatus.FORBIDDEN,"FILE_READ_FORBIDDEN","Không có quyền đọc tệp");
        }
        return e;
    }
    private ResponseEntity<Resource> resource(StoredFileEntity e){
        FileSystemResource r=new FileSystemResource(path(e.storageKey));
        return ResponseEntity.ok().contentType(MediaType.parseMediaType(e.contentType)).contentLength(e.sizeBytes).header(HttpHeaders.CONTENT_DISPOSITION,ContentDisposition.attachment().filename(e.originalName,StandardCharsets.UTF_8).build().toString()).body(r);
    }
    private Path path(String key){
        Path p=root.resolve(key).normalize();
        if(!p.startsWith(root))bad("INVALID_PATH","Đường dẫn không hợp lệ");
        return p;
    }
    private String safe(String n){
        String s=n==null?"upload.bin":Paths.get(n).getFileName().toString();
        return s.replaceAll("[\\r\\n\\0]","_").substring(0,Math.min(240,s.length()));
    }
    private String sha(byte[] b){
        try{
            return HexFormat.of().formatHex(MessageDigest.getInstance("SHA-256").digest(b));
        }
        catch(Exception e){
            throw new IllegalStateException(e);
        }

    }
    private StoredFileResponse view(StoredFileEntity e){
        return new StoredFileResponse(e.id,e.originalName,e.contentType,e.sizeBytes,e.sha256,e.purpose,e.status,e.createdAt);
    }
    private UUID user(){
        try{
            return CurrentUser.id();
        }
        catch(Exception e){
            return new UUID(0,1);
        }

    }
    private void bad(String c,String m){
        throw new ApiException(HttpStatus.BAD_REQUEST,c,m);
    }

}
@RestController
@RequestMapping("/api/v1/files")
public class FileStorageApi {
    private final FileStorageService service;
    public FileStorageApi(FileStorageService s){
        service=s;
    }
    @GetMapping
    public List<StoredFileResponse> list(){
        return service.list();
    }
    @PostMapping(consumes=MediaType.MULTIPART_FORM_DATA_VALUE)
    @ResponseStatus(HttpStatus.CREATED)
    public StoredFileResponse upload(@RequestPart("file")MultipartFile f,@RequestParam(defaultValue="GENERAL")String purpose){
        return service.store(f,purpose);
    }
    @GetMapping("/{id}")
    public StoredFileResponse metadata(@PathVariable UUID id){
        return service.metadata(id);
    }
    @GetMapping("/{id}/docx-preview")
    public DocumentPreviewResponse preview(@PathVariable UUID id){
        return service.preview(id);
    }
    @GetMapping("/{id}/content")
    public ResponseEntity<Resource> content(@PathVariable UUID id,@RequestHeader(value="Range",required=false)String range){
        return service.content(id,range);
    }
    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void delete(@PathVariable UUID id){
        service.delete(id);
    }

}
@RestController
@RequestMapping("/internal/v1/files")
class InternalFileController {
    private final FileStorageService service;
    private final com.lmspilot.support.security.InternalTokenAuthorizer token;
    InternalFileController(FileStorageService s,com.lmspilot.support.security.InternalTokenAuthorizer t){
        service=s;
        token=t;
    }
    @GetMapping("/{id}/content") ResponseEntity<Resource> content(@PathVariable UUID id,@RequestHeader(value="X-Service-Token",required=false)String k){
        token.require(k);
        return service.internalContent(id);
    }
    @GetMapping("/{id}")
    InternalStoredFileResponse metadata(@PathVariable UUID id,@RequestHeader(value="X-Service-Token",required=false)String k){
        token.require(k);
        return service.internalMetadata(id);
    }
    @PostMapping("/access-grants")
    void grant(@RequestBody GrantFileAccessRequest i,@RequestHeader(value="X-Service-Token",required=false)String k){
        token.require(k);
        service.grant(i);
    }

}

File: backend/services/file-storage-service/src/main/kotlin/com/lmspilot/filestorage/api/FileStorageApi.kt
package com.lmspilot.filestorage.api

import com.lmspilot.contracts.Permissions
import com.lmspilot.filestorage.domain.StoredFileEntity
import com.lmspilot.filestorage.domain.StoredFileRepository
import com.lmspilot.filestorage.domain.StoredFileStatus
import com.lmspilot.filestorage.domain.FileAccessGrantEntity
import com.lmspilot.filestorage.domain.FileAccessGrantRepository
import com.lmspilot.support.api.ApiException
import com.lmspilot.support.security.CurrentUser
import jakarta.annotation.PostConstruct
import org.springframework.beans.factory.annotation.Value
import org.springframework.core.io.FileSystemResource
import org.springframework.core.io.support.ResourceRegion
import org.springframework.http.ContentDisposition
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpRange
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestHeader
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RequestPart
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.multipart.MultipartFile
import java.nio.file.AtomicMoveNotSupportedException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.nio.file.StandardCopyOption
import java.nio.file.StandardOpenOption
import java.security.MessageDigest
import java.time.Instant
import java.util.Locale
import java.util.UUID

private const val MAX_PURPOSE_LENGTH = 50

data class StoredFileResponse(
    val id: UUID,
    val originalName: String,
    val contentType: String,
    val sizeBytes: Long,
    val sha256: String,
    val purpose: String,
    val status: StoredFileStatus,
    val createdAt: Instant,
)

data class InternalStoredFileResponse(
    val id: UUID,
    val ownerId: UUID,
    val originalName: String,
    val contentType: String,
    val sizeBytes: Long,
    val sha256: String,
    val purpose: String,
    val status: StoredFileStatus,
    val createdAt: Instant,
)

@Service
class FileStorageService(
    private val repository: StoredFileRepository,
    private val grants: FileAccessGrantRepository,
    @Value("\${storage.root:./data/files}") root: String,
    @Value("\${storage.max-size-bytes:209715200}") private val maxSizeBytes: Long,
    @Value("\${storage.allowed-content-types:application/pdf,text/plain}") allowedContentTypes: String,
) {
    private val rootPath = Paths.get(root).toAbsolutePath().normalize()
    private val blockedExtensions = setOf("exe", "dll", "bat", "cmd", "ps1", "sh", "jar", "msi", "com", "scr", "js", "html", "htm", "svg")
    private val allowedTypes = allowedContentTypes.split(',').map(String::trim).filter(String::isNotBlank).map(String::lowercase).toSet()
    private val extensionTypes = mapOf(
        "pdf" to "application/pdf",
        "docx" to "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        "mp4" to "video/mp4",
        "mp3" to "audio/mpeg",
        "png" to "image/png",
        "jpg" to "image/jpeg",
        "jpeg" to "image/jpeg",
        "txt" to "text/plain",
        "csv" to "text/csv",
    )

    @PostConstruct
    fun initialize() {
        require(maxSizeBytes > 0) { "FILE_MAX_SIZE_BYTES must be positive" }
        require(allowedTypes.isNotEmpty()) { "FILE_ALLOWED_TYPES must not be empty" }
        Files.createDirectories(rootPath)
    }

    @Transactional
    fun store(file: MultipartFile, purposeInput: String): StoredFileResponse {
        if (file.isEmpty) throw ApiException(HttpStatus.BAD_REQUEST, "FILE_EMPTY", "Tệp tải lên đang trống")
        if (file.size > maxSizeBytes) throw ApiException(HttpStatus.PAYLOAD_TOO_LARGE, "FILE_TOO_LARGE", "Tệp vượt quá dung lượng cho phép")

        val safeName = safeFilename(file.originalFilename)
        val extension = safeName.substringAfterLast('.', "").lowercase(Locale.ROOT)
        if (extension.isBlank() || extension in blockedExtensions) {
            throw ApiException(HttpStatus.UNSUPPORTED_MEDIA_TYPE, "FILE_TYPE_BLOCKED", "Loại tệp không được phép")
        }
        val contentType = normalizedContentType(file.contentType, extension)
        if (contentType !in allowedTypes) {
            throw ApiException(HttpStatus.UNSUPPORTED_MEDIA_TYPE, "FILE_TYPE_NOT_ALLOWED", "Định dạng tệp không nằm trong danh sách cho phép")
        }
        val purpose = normalizePurpose(purposeInput)

        val id = UUID.randomUUID()
        val storageKey = "${id.toString().substring(0, 2)}/$id"
        val target = checkedPath(storageKey)
        Files.createDirectories(target.parent)
        val temporary = Files.createTempFile(target.parent, ".upload-", ".tmp")
        val digest = MessageDigest.getInstance("SHA-256")

        try {
            file.inputStream.use { input ->
                Files.newOutputStream(temporary, StandardOpenOption.TRUNCATE_EXISTING).use { output ->
                    val buffer = ByteArray(DEFAULT_BUFFER_SIZE)
                    var written = 0L
                    while (true) {
                        val read = input.read(buffer)
                        if (read < 0) break
                        written += read
                        if (written > maxSizeBytes) {
                            throw ApiException(HttpStatus.PAYLOAD_TOO_LARGE, "FILE_TOO_LARGE", "Tệp vượt quá dung lượng cho phép")
                        }
                        digest.update(buffer, 0, read)
                        output.write(buffer, 0, read)
                    }
                    if (written != file.size) {
                        throw ApiException(HttpStatus.BAD_REQUEST, "FILE_SIZE_MISMATCH", "Dung lượng tệp tải lên không hợp lệ")
                    }
                }
            }
            moveAtomically(temporary, target)
            return try {
                repository.save(
                    StoredFileEntity(
                        id = id,
                        ownerId = CurrentUser.id(),
                        originalName = safeName,
                        storageKey = storageKey,
                        contentType = contentType,
                        sizeBytes = file.size,
                        sha256 = digest.digest().joinToString("") { "%02x".format(it) },
                        purpose = purpose,
                    )
                ).response()
            } catch (ex: Exception) {
                Files.deleteIfExists(target)
                throw ex
            }
        } finally {
            Files.deleteIfExists(temporary)
        }
    }

    @Transactional(readOnly = true)
    fun list(): List<StoredFileResponse> = (if (Permissions.OPERATIONS_MANAGE in CurrentUser.authorities()) repository.findAll() else repository.findAllByOwnerIdOrderByCreatedAtDesc(CurrentUser.id()))
        .filter { it.status == StoredFileStatus.AVAILABLE }
        .sortedByDescending { it.createdAt }
        .map { it.response() }

    @Transactional(readOnly = true)
    fun metadata(id: UUID): StoredFileResponse = readable(id).response()

    @Transactional(readOnly = true)
    fun internalMetadata(id: UUID): InternalStoredFileResponse = available(id).internalResponse()

    @Transactional(readOnly = true)
    fun download(id: UUID, inline: Boolean = false, rangeHeader: String? = null): ResponseEntity<*> {
        val entity = readable(id)
        return fileResponse(entity, inline, rangeHeader)
    }

    @Transactional(readOnly = true)
    fun internalDownload(id: UUID, inline: Boolean = false): ResponseEntity<*> =
        fileResponse(available(id), inline)

    @Transactional
    fun grantAccess(userId: UUID, fileIds: Set<UUID>, sourceInput: String, ttlSeconds: Long) {
        if (fileIds.size > 500) throw ApiException(HttpStatus.BAD_REQUEST, "FILE_GRANT_LIMIT", "Mỗi yêu cầu chỉ được cấp tối đa 500 tệp")
        val source = sourceInput.trim().uppercase(Locale.ROOT)
        if (source.isBlank() || source.length > 80 || !source.matches(Regex("[A-Z0-9_:-]+"))) {
            throw ApiException(HttpStatus.BAD_REQUEST, "INVALID_GRANT_SOURCE", "Nguồn cấp quyền tệp không hợp lệ")
        }
        val now = Instant.now()
        val expiresAt = now.plusSeconds(ttlSeconds.coerceIn(60, 14_400))
        fileIds.forEach { fileId ->
            val file = available(fileId)
            if (file.purpose !in setOf("COURSE_CONTENT", "ASSIGNMENT_SUBMISSION", "COURSE_DOCUMENT", "NEWS_ATTACHMENT")) {
                throw ApiException(HttpStatus.BAD_REQUEST, "FILE_PURPOSE_NOT_GRANTABLE", "Không thể cấp quyền chia sẻ cho mục đích tệp này")
            }
            val grant = grants.findByFileIdAndUserId(fileId, userId)
                ?: FileAccessGrantEntity(fileId = fileId, userId = userId, createdAt = now)
            grant.source = source
            grant.expiresAt = expiresAt
            grant.updatedAt = now
            grants.save(grant)
        }
    }

    private fun fileResponse(entity: StoredFileEntity, inline: Boolean, rangeHeader: String? = null): ResponseEntity<*> {
        val path = checkedPath(entity.storageKey)
        if (!Files.isRegularFile(path)) throw ApiException(HttpStatus.NOT_FOUND, "FILE_BYTES_MISSING", "Dữ liệu tệp không còn tồn tại")
        val resource = FileSystemResource(path)
        val mediaType = runCatching { MediaType.parseMediaType(entity.contentType) }.getOrDefault(MediaType.APPLICATION_OCTET_STREAM)
        val disposition = (if (inline) ContentDisposition.inline() else ContentDisposition.attachment()).filename(entity.originalName, Charsets.UTF_8).build().toString()
        val region = requestedRegion(rangeHeader, resource, entity.sizeBytes)
        if (region != null) {
            return ResponseEntity.status(HttpStatus.PARTIAL_CONTENT)
                .contentType(mediaType)
                .contentLength(region.count)
                .header(HttpHeaders.CONTENT_DISPOSITION, disposition)
                .header(HttpHeaders.ACCEPT_RANGES, "bytes")
                .header("X-Content-Type-Options", "nosniff")
                .body(region)
        }
        return ResponseEntity.ok()
            .contentType(mediaType)
            .contentLength(entity.sizeBytes)
            .header(HttpHeaders.CONTENT_DISPOSITION, disposition)
            .header(HttpHeaders.ACCEPT_RANGES, "bytes")
            .header("X-Content-Type-Options", "nosniff")
            .body(resource)
    }

    private fun requestedRegion(rangeHeader: String?, resource: FileSystemResource, size: Long): ResourceRegion? {
        if (rangeHeader.isNullOrBlank()) return null
        return try {
            val range = HttpRange.parseRanges(rangeHeader).singleOrNull()
                ?: throw IllegalArgumentException("Only one byte range is supported")
            val start = range.getRangeStart(size)
            val end = range.getRangeEnd(size)
            if (start !in 0 until size || end < start) throw IllegalArgumentException("Unsatisfiable byte range")
            ResourceRegion(resource, start, end - start + 1)
        } catch (_: IllegalArgumentException) {
            throw ApiException(HttpStatus.REQUESTED_RANGE_NOT_SATISFIABLE, "INVALID_FILE_RANGE", "Khoảng byte được yêu cầu không hợp lệ")
        }
    }

    @Transactional
    fun delete(id: UUID) {
        val entity = available(id)
        if (entity.ownerId != CurrentUser.id() && Permissions.OPERATIONS_MANAGE !in CurrentUser.authorities()) {
            throw ApiException(HttpStatus.FORBIDDEN, "FILE_OWNER_MISMATCH", "Bạn không có quyền xóa tệp")
        }
        entity.status = StoredFileStatus.DELETED
        entity.deletedAt = Instant.now()
    }

    private fun available(id: UUID): StoredFileEntity {
        val entity = repository.findById(id).orElseThrow { ApiException(HttpStatus.NOT_FOUND, "FILE_NOT_FOUND", "Không tìm thấy tệp") }
        if (entity.status != StoredFileStatus.AVAILABLE) throw ApiException(HttpStatus.GONE, "FILE_UNAVAILABLE", "Tệp không còn khả dụng")
        return entity
    }

    /**
     * Public file routes must enforce object access in addition to the coarse
     * FILES_DOWNLOAD authority carried by the JWT. Internal service routes use
     * [available] after validating their service token.
     */
    private fun readable(id: UUID): StoredFileEntity {
        val entity = available(id)
        if (entity.ownerId == CurrentUser.id() || hasAdministrativeFileAccess()) return entity

        if (grants.findByFileIdAndUserId(entity.id, CurrentUser.id())?.expiresAt?.isAfter(Instant.now()) == true) return entity

        throw ApiException(HttpStatus.FORBIDDEN, "FILE_READ_FORBIDDEN", "Bạn không có quyền đọc tệp này")
    }

    private fun hasAdministrativeFileAccess(): Boolean =
        CurrentUser.isSystemAdmin() ||
            "ADMIN" in CurrentUser.roles() ||
            Permissions.OPERATIONS_MANAGE in CurrentUser.authorities()

    private fun checkedPath(storageKey: String): Path {
        val path = rootPath.resolve(storageKey).normalize()
        if (!path.startsWith(rootPath)) throw ApiException(HttpStatus.BAD_REQUEST, "INVALID_PATH", "Đường dẫn tệp không hợp lệ")
        return path
    }

    private fun safeFilename(value: String?): String {
        val normalized = (value ?: "file.bin").replace('\\', '/').substringAfterLast('/').trim()
        if (normalized.isBlank() || normalized.length > 255 || normalized.any { it.code < 32 }) {
            throw ApiException(HttpStatus.BAD_REQUEST, "INVALID_FILENAME", "Tên tệp không hợp lệ")
        }
        return normalized
    }

    private fun normalizedContentType(value: String?, extension: String): String {
        val declared = value?.substringBefore(';')?.trim()?.lowercase(Locale.ROOT).orEmpty()
        if (declared.isBlank() || declared == MediaType.APPLICATION_OCTET_STREAM_VALUE) {
            return extensionTypes[extension] ?: declared.ifBlank { MediaType.APPLICATION_OCTET_STREAM_VALUE }
        }
        val expected = extensionTypes[extension]
        if (expected != null && declared != expected) {
            throw ApiException(HttpStatus.UNSUPPORTED_MEDIA_TYPE, "FILE_TYPE_MISMATCH", "Phần mở rộng và loại nội dung của tệp không khớp")
        }
        return declared
    }

    private fun normalizePurpose(value: String): String {
        val purpose = value.trim().uppercase(Locale.ROOT)
        if (purpose.isBlank() || purpose.length > MAX_PURPOSE_LENGTH || !purpose.matches(Regex("[A-Z0-9_]+"))) {
            throw ApiException(HttpStatus.BAD_REQUEST, "INVALID_FILE_PURPOSE", "Mục đích sử dụng tệp không hợp lệ")
        }
        return purpose
    }

    private fun moveAtomically(source: Path, target: Path) {
        try {
            Files.move(source, target, StandardCopyOption.ATOMIC_MOVE)
        } catch (_: AtomicMoveNotSupportedException) {
            Files.move(source, target)
        }
    }
}

private fun StoredFileEntity.response() = StoredFileResponse(id, originalName, contentType, sizeBytes, sha256, purpose, status, createdAt)
private fun StoredFileEntity.internalResponse() = InternalStoredFileResponse(id, ownerId, originalName, contentType, sizeBytes, sha256, purpose, status, createdAt)

@RestController
@RequestMapping("/api/v1/files")
class FileController(private val service: FileStorageService) {
    @GetMapping
    @PreAuthorize("hasAnyAuthority('${Permissions.FILES_READ}','${Permissions.FILES_DOWNLOAD}')")
    fun list() = service.list()

    @PostMapping(consumes = [MediaType.MULTIPART_FORM_DATA_VALUE])
    @PreAuthorize("hasAuthority('${Permissions.FILES_UPLOAD}')")
    fun upload(@RequestPart("file") file: MultipartFile, @RequestParam(defaultValue = "GENERAL") purpose: String) = service.store(file, purpose)

    @GetMapping("/{id}")
    @PreAuthorize("hasAuthority('${Permissions.FILES_DOWNLOAD}')")
    fun metadata(@PathVariable id: UUID) = service.metadata(id)

    @GetMapping("/{id}/content")
    @PreAuthorize("hasAuthority('${Permissions.FILES_DOWNLOAD}')")
    fun download(
        @PathVariable id: UUID,
        @RequestParam(defaultValue = "false") inline: Boolean,
        @RequestHeader(HttpHeaders.RANGE, required = false) range: String?,
    ) = service.download(id, inline, range)

    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    @PreAuthorize("isAuthenticated()")
    fun delete(@PathVariable id: UUID) = service.delete(id)
}
